mls-rs/src/group/message_processor.rs

```diff
@@ -776,34 +630,38 @@ pub(crate) trait MessageProcessor: Send + Sync {
 
         let sender = commit_sender(&auth_content.content.sender, &provisional_state)?;
 
-        #[cfg(feature = "state_update")]
-        let mut state_update = self
-            .make_state_update(&provisional_state, commit.path.as_ref(), sender)
-            .await?;
-
-        #[cfg(not(feature = "state_update"))]
-        let state_update = StateUpdate {};
-
         //Verify that the path value is populated if the proposals vector contains any Update
         // or Remove proposals, or if it's empty. Otherwise, the path value MAY be omitted.
         if path_update_required(&provisional_state.applied_proposals) && commit.path.is_none() {
             return Err(MlsError::CommitMissingPath);
         }
 
         if !self.can_continue_processing(&provisional_state) {
-            #[cfg(feature = "state_update")]
-            {
-                state_update.active = false;
-            }
-
             return Ok(CommitMessageDescription {
                 is_external: matches!(auth_content.content.sender, Sender::NewMemberCommit),
                 authenticated_data: auth_content.content.authenticated_data,
                 committer: *sender,
-                state_update,
+                effect: CommitEffect::Removed,
             });
         }
 
+        let commit_effect =
+            if let Some(reinit) = provisional_state.applied_proposals.reinitializations.pop() {
+                self.group_state_mut().pending_reinit = Some(reinit.proposal.clone());
+                CommitEffect::ReInit(reinit)
+            } else {
+                CommitEffect::NewEpoch(Box::new(NewEpoch {
+                    epoch: provisional_state.group_context.epoch,
+                    prior_state: self.group_state().clone(),
```

Comment:
Intuitively, we should be able to move out from the current state instead of clone, since we're going to replace it by the provisional one? We're cloning a large thing here. Can be a follow-up, maybe it's a bigger change.

Reply:
It would definitely be a larger change since we don't make the swap until later when every succeeds. If you do a swap here and then there is an error between that and the end of processing you would wind up with a corrupted state

Reply:
Yes, I was thinking that we could set this value as the very last thing, but that's hard with the way `update_key_schedule` makes the final changes, and it's a trait function... Would be nice to have at some point, maybe?